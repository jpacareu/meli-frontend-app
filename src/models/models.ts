import { History } from 'history';

export interface IState {
	items: IItem[];
	searchBox: string;
	categories: string[];
	loadingSearch: boolean;
	detailProduct: {
		item: IDetailItem;
	} | null;
}

export interface IActions {
	search: (...rest: any) => {};
	getProductDetail: (...rest: any) => {};
	loading: (...rest: any) => {};
	received: (...rest: any) => {};
	setState: (...rest: any) => {};
	handleError: (...rest: any) => {};
}

export interface IStateHistory extends IState {
	history: History;
}

export interface IDetailItem {
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
	sold_quantity: number;
	description: string;
}

export interface IItem {
	id: string;
	title: string;
	price: IPrice;
	picture: string;
	condition: ICondition;
	free_shipping: boolean;
}

export type ICondition = 'new' | 'used';

export interface IPrice {
	currency: string;
	amount: number;
	decimals: number;
}

export interface IStore {
	Provider: any;
	connect: () => any;
	actions: IActions;
}
