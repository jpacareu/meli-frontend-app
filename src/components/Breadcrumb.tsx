import React from 'react';

interface IProps {
	list: string[];
}

export const Breadcrumb = (props: IProps) => {
	const list: string[] = props.list || [];
	return (
		<nav aria-label="breadcrumb">
			<ol className="breadcrumb-category d-md-flex">
				{list.map((el, i, arr) => {
					const isStrong = i === arr.length - 1 ? '__highlight' : '';
					return (
						<li
							key={i}
							className={`breadcrumb-category__item${isStrong}`}
							aria-current="page">
							{el}
						</li>
					);
				})}
			</ol>
		</nav>
	);
};
