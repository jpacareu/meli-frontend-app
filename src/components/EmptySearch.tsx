export const EmptySearch = () => {
	return null;
};
