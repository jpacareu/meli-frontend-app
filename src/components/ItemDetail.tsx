import React from 'react';
import { IDetailItem, ICondition, IPrice } from '../models/models';
import { formatAmount, formatCurrency } from '../utils/money';

interface IProps {
	detail: IDetailItem;
}

export const ItemDetail = (props: IProps) => {
	const { detail } = props;

	return (
		<div className="item-detail row">
			<div className="col-md-7">
				<Image title={detail.title} picture={detail.picture} />
			</div>
			<div className="col-md-4 col-sm-12">
				<section className="item-detail__main-info">
					<div className="item-detail__condition-and-sold d-flex">
						<Condition condition={detail.condition} />
						<span> - </span>
						<Sold sold={detail.sold_quantity} />
					</div>
					<Title title={detail.title} />
					<Price price={detail.price} />
					<ButtonComprar />
				</section>
			</div>
			<div className="col-md-7 col-sm-12">
				<Detalle description={detail.description} />
			</div>
		</div>
	);
};

export const Image = (props: { title: string; picture: string }) => {
	return (
		<img className="item-detail__image" alt={props.title} src={props.picture} />
	);
};
export const Detalle = (props: { description: string }) => {
	return (
		<section>
			<summary className="item-detail__descripcion-label">
				Descripci√≥n del producto
			</summary>
			<div className="item-detail__descripcion-text">{props.description}</div>
		</section>
	);
};

export const Condition = (props: { condition: ICondition }) => {
	return (
		<div className="item-detail__condition">
			{props.condition === 'new' ? 'Nuevo' : 'Usado'}
		</div>
	);
};

export const Sold = (props: { sold: number }) => {
	return <div className="item-detail__sold">{props.sold}</div>;
};

export const Title = (props: { title: string }) => {
	return <div className="item-detail__title">{props.title}</div>;
};

export const Price = (props: { price: IPrice }) => {
	return (
		<div className="item-detail__price">{`${formatCurrency(
			props.price.currency
		)} ${formatAmount(props.price.amount)}`}</div>
	);
};

export const ButtonComprar = () => {
	return <button className="item-detail__button">Comprar</button>;
};
