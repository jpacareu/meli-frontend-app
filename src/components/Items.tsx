import React from 'react';
import { IStateHistory, IItem, IPrice } from '../models/models';
import freeShippingLogo from '../assets/ic_shipping.png';
import { formatAmount, formatCurrency } from '../utils/money';
type PushFunction = (location: { pathname: string }) => void;

interface IItemWithRouter {
	push: PushFunction;
	model: IItem;
}

export const Items = (props: IStateHistory) => {
	return (
		<div className="result-search">
			{props.items.map((elem, index) => (
				<Item key={index} push={props.history.push} model={elem} />
			))}
		</div>
	);
};

const goProductDetail = (id: string, push: PushFunction) => {
	return () =>
		push({
			pathname: `/items/${id}`
		});
};

export const Item = (props: IItemWithRouter) => {
	const { model, push } = props;

	return (
		<div className="col-md-12">
			<div className="item">
				<Image model={model} push={push} />
				<div>
					<div className="d-flex flex-row">
						<div className="item__price d-flex align-items-center">
							<Price price={model.price} />
							{model.free_shipping && <FreeShipping />}
						</div>
					</div>
					<Title title={model.title} />
				</div>
			</div>
		</div>
	);
};

const Title = (props: { title: string }) => (
	<p className="item__title">{props.title}</p>
);

export const FreeShipping = () => {
	return (
		<img
			className="item__free_shipping"
			alt="Free shipping"
			src={freeShippingLogo}
		/>
	);
};

const Image = (props: { model: IItem; push: PushFunction }) => {
	return (
		<img
			alt={props.model.title}
			className="item__image"
			onClick={goProductDetail(props.model.id, props.push)}
			src={props.model.picture}
		/>
	);
};

const Price = (props: { price: IPrice }) => {
	return (
		<div>{`${formatCurrency(props.price.currency)} ${formatAmount(
			props.price.amount
		)}`}</div>
	);
};
