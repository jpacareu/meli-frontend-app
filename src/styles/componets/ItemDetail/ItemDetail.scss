.item-detail {
    background-color: $white;
    &__image {
        max-height: 680px;
        width: 100%;
        height: 100%;
        border-radius: 4px;
    }
    &__main-info {
        padding: $padding-xs;
    }
    &__condition-and-sold {
        margin-top: $padding-lg;
        margin-bottom: $padding-sm;
        font-size: $font-size-14; 
        & > * {
            margin-right: 10px;
        }
    }
    &__title {
        font-size: $font-size-24;
        font-weight: 600;
        margin-bottom: $padding-lg;
    }
    &__price {
        font-size: $font-size-46;
        font-weight: 300;
        margin-bottom: $padding-lg;
    }
    &__button {
        background-color: $primary;
        color: $white;
        border-radius: 5px;
        width: 80%;
        height: 40px;
    }
    &__descripcion-label {
        color: $dark-gray;
        font-size: $font-size-28;
        font-weight: 500;
        margin-left: $padding-sm;
    }
    &__descripcion-text {
        color: $light-gray;
        font-size: $font-size-14;
        margin: $padding-sm 0 $padding-sm $padding-sm;
    }
}