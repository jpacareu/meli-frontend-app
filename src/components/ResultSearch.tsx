import React, { Fragment } from 'react';
import { connect } from '../store';
import Spinner from './Spinner';
import { IState, IStateHistory } from '../models/models';
import { Items } from './Items';
import { Breadcrumb } from './Breadcrumb';

export let ResultSearch = (props: IStateHistory) => {
	return props.loadingSearch ? (
		<Spinner />
	) : (
		<Fragment>
			<Breadcrumb list={props.categories} />
			<Items {...props} />
		</Fragment>
	);
};

export default ResultSearch = connect((state: IState) => ({ ...state }))(
	ResultSearch
);
