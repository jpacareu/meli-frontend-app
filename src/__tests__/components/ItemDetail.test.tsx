import React from 'react';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { ItemDetail, Condition } from '../../components/ItemDetail';
import { IDetailItem } from '../../models/models';

const detail: IDetailItem = {
	condition: 'new',
	description: 'This is the desciption',
	picture: 'http://mypicture.com/image.jpg',
	sold_quantity: 1000,
	title: 'Title 1',
	price: {
		amount: 2500,
		currency: 'ARS',
		decimals: 0
	},
	free_shipping: true
};

let component: JSX.Element | null = null;
let tree: ReactTestRenderer;

describe('<ItemDetail />', () => {
	beforeEach(() => {
		component = <ItemDetail detail={detail} />;
		tree = renderer.create(component);
	});
	it('should render without crashing', () => {
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("should render 'Usado'", () => {
		detail.condition = 'used';

		component = <ItemDetail detail={detail} />;
		tree = renderer.create(component);

		const condition = tree.root.findByType(Condition);
		const children = condition.findByType('div').children;
		expect(children[0]).toBe('Usado');
	});
});
