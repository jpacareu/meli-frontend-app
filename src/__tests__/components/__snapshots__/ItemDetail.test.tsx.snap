// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemDetail /> should render without crashing 1`] = `
<div
  className="item-detail row"
>
  <div
    className="col-md-7"
  >
    <img
      alt="Title 1"
      className="item-detail__image"
      src="http://mypicture.com/image.jpg"
    />
  </div>
  <div
    className="col-md-4 col-sm-12"
  >
    <section
      className="item-detail__main-info"
    >
      <div
        className="item-detail__condition-and-sold d-flex"
      >
        <div
          className="item-detail__condition"
        >
          Nuevo
        </div>
        <span>
           - 
        </span>
        <div
          className="item-detail__sold"
        >
          1000
        </div>
      </div>
      <div
        className="item-detail__title"
      >
        Title 1
      </div>
      <div
        className="item-detail__price"
      >
        $ 2.500
      </div>
      <button
        className="item-detail__button"
      >
        Comprar
      </button>
    </section>
  </div>
  <div
    className="col-md-7 col-sm-12"
  >
    <section>
      <summary
        className="item-detail__descripcion-label"
      >
        Descripci√≥n del producto
      </summary>
      <div
        className="item-detail__descripcion-text"
      >
        This is the desciption
      </div>
    </section>
  </div>
</div>
`;
